net/retriever/retriever.go

```diff
@@ -92,10 +93,16 @@ func (m *MessageRetriever) fetchPointers() {
 	// Wait for each goroutine to finish then process any remaining messages that needed
 	// to be processed last
 	wg.Wait()
-	for _, env := range m.messageQueue {
-		m.handleMessage(env, nil)
+
+DRAIN_LOOP:
+	for {
+		select {
+		case env := <-m.messageQueue:
+			m.handleMessage(env, nil)
+		default:
+			break DRAIN_LOOP
```

Comment:
It executes if and only if none of the other cases can be fulfilled. In this case that means that the m.messageQueue channel is empty. So the entire loop construction is pull items off the queue and handling them until the queue is empty. Once it's empty it breaks from the loop and continues on.